
syntax = "proto3";

package ngraph.runtime.he.pb;

/// \brief Represents a message between the server and client.
message TCPMessage {
  enum Type { UNKNOWN = 0; REQUEST = 1; RESPONSE = 2; }
  Type type = 1;
  Function function = 2;
  EncryptionParameters encryption_parameters = 3;
  EvaluationKey eval_key = 4;
  PublicKey public_key = 5;
  repeated HETensor he_tensors = 6;
}

message EncryptionParameters { bytes encryption_parameters = 1; }

message EvaluationKey { bytes eval_key = 1; }

message PublicKey { bytes public_key = 1; }

message Function { string function = 1; }

message HETensor {
  string name = 1;
  repeated uint64 shape = 2;
  enum ElementType {
    UNDEFINED = 0; DYNAMIC = 1; BOOLEAN = 2; BF16 = 3; F16 = 4; F32 = 5;
    F64 = 6;
    I8 = 7;
    I16 = 8;
    I32 = 9;
    I64 = 10;
    U1 = 11;
    U8 = 12;
    U16 = 13;
    U32 = 14;
    U64 = 15;
  }
  ElementType type = 3;
  bool packed = 4;
  uint64 offset = 5;
  repeated HEType data = 6;
}

message HEType {
  bool is_plaintext = 1;
  bool plaintext_packing = 2;
  bool complex_packing = 3;
  uint64 batch_size = 4;
  repeated float plain = 5;
  bytes ciphertext = 6;
}
